source/common/router/router.cc

```diff
@@ -1470,6 +1470,8 @@ void Filter::onUpstreamHeaders(uint64_t response_code, Http::ResponseHeaderMapPt
   for (const auto& header_value : downstream_set_cookies_) {
     headers->addReferenceKey(Http::Headers::get().SetCookie, header_value);
   }
+  
+  callbacks_->streamInfo().setResponseCodeDetails(StreamInfo::ResponseCodeDetails::get().ViaUpstream);
```

Comment:
so below I see callbacks_->encodeHeaders(.. ViaUpstream).  Is that not sufficient? Perhaps you could add a regression test which fails with the current code
/wait

Reply:
hm, can you tell me a bit more about your manual testing?
I wouldn't expect the details to be set in unit tests, say, because the function which sets it would be mocked out.
We do have a number of integration tests which verify "via_upstream" makes it into access logs for the happy path (grep -r via_upstream test/integration/*) so if you can tell me a bit more about where you're not seeing it, maybe I can help sort out what's going on!

Reply:
> so below I see callbacks_->encodeHeaders(.. ViaUpstream).  Is that not sufficient? Perhaps you could add a regression test which fails with the current code
> /wait

The finalizeRespnseHeaders will called before the setRespnseCodeDetail, which results in #

Reply:
@alyssawilk Yes, as I mentioned in the associated issue, the access logs work fine, it's only the response headers that fail. It did feel like an ordering issue to me, as @wbpcode has mentioned, but I was unsure of how to address that more specifically.

Reply:
No the fix is totally fine (sorry for missing that linked bug) but you do need a test.
given it's likely my bug, I've whipped one up over here: https://github.com/envoyproxy/envoy/compare/main...alyssawilk:add?expand=1 
you can either copy the test / release notes and put them in your PR, or I can create a PR for that one and just co-author attribute you.

Reply:
Fantastic and thank you! I've copied your test and will add to my PR.

Reply:
mind copying the release notes over too?  worth noting functional changes for this sort of thing :-)

Reply:
alas I'm not sure what's going on locally. maybe you didn't git add and commit that file?
/wait on CI

Reply:
It was really strange, it just wasn't showing up at all... but I redid the edit and it worked this time. Thanks again!